src/dpl/src/Grid.cpp

```diff
@@ -322,20 +469,27 @@ void Opendp::groupAssignCellRegions()
 
 void Opendp::groupInitPixels2()
 {
-  for (int x = 0; x < row_site_count_; x++) {
-    for (int y = 0; y < row_count_; y++) {
-      Rect sub;
-      sub.init(x * site_width_,
-               y * row_height_,
-               (x + 1) * site_width_,
-               (y + 1) * row_height_);
-      Pixel* pixel = gridPixel(x, y);
-      for (Group& group : groups_) {
-        for (Rect& rect : group.regions) {
-          if (!isInside(sub, rect) && checkOverlap(sub, rect)) {
-            pixel->util = 0.0;
-            pixel->cell = &dummy_cell_;
-            pixel->is_valid = false;
+  for (auto& layer : grid_layers_) {
+    int row_height = layer.first;
+    LayerInfo& layer_info = layer.second;
+    int row_count = divFloor(core_.dy(), row_height);
+    int row_site_count = divFloor(core_.dx(), site_width_);
+    for (int x = 0; x < row_site_count; x++) {
+      for (int y = 0; y < row_count; y++) {
+        Rect sub;
+        // TODO: Site width here is wrong if multiple site widths are supported!
```

Comment:
I don't have a strong opinion on that. Currently, we only support the same site width in DPL. In my PR I added support to multi-height cells using the new 3D grid. It might be work for another PR to support multiple site-widths in DPL because it is tricky. 

Each row only has 1 site-width, but rows can overlap, and from what I imagine, supporting different site widths might require a separate grid for each unique site-row and site-width. Do possible site-widths have to be multiples of each other ? if yes, then the solution might be simpler.

Reply:
I have not personally seen any PDKs where multi-height cells use different site widths. I think it is fine to assume each overlapping row has the same site width.

A block can have different rows with different sites, but typically these don't overlap e.g. for IO rows.